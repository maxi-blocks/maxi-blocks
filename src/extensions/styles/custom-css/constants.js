const PSEUDO_ELEMENTS = ['before', 'after'];

export default PSEUDO_ELEMENTS;
