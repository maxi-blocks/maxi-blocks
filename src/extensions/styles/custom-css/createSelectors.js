const createSelectors = (selectors, addPseudoElementSelectors = true) => {
	const getNormalAndHoverSelectors = obj => {
		const { label, target } = obj;

		return {
			normal: obj,
			hover: {
				label: `${label} on hover`,
				target: `${target}:hover`,
			},
		};
	};

	const addPseudoElementSelector = (key, selector, pseudoElement, obj) => {
		const { label, target } = selector;

		obj[`${pseudoElement} ${key}`] = {
			label: `${label} ::${pseudoElement}`,
			target: `${target}::${pseudoElement}`,
		};
	};

	const result = { ...selectors };

	if (addPseudoElementSelectors)
		['before', 'after'].forEach(pseudoElement => {
			Object.entries(selectors).forEach(([key, selector]) => {
				addPseudoElementSelector(key, selector, pseudoElement, result);
			});
		});

	Object.entries(result).forEach(([key, obj]) => {
		result[key] = getNormalAndHoverSelectors(obj);
	});

	return result;
};

export default createSelectors;
