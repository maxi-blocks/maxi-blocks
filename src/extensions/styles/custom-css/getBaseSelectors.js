const getBaseSelectors = selectors =>
	Object.fromEntries(
		Object.entries(selectors).map(([label, target]) => [
			label,
			{ label, target },
		])
	);

export default getBaseSelectors;
