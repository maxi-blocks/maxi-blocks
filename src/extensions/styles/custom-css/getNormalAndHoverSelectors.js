import PSEUDO_ELEMENTS from './constants';

const getNormalAndHoverSelectors = ({ label, target }) => {
	const pseudoElement = PSEUDO_ELEMENTS.find(pseudo =>
		label.includes(pseudo)
	);
	const targetWithoutPseudoElement = target.replace(`::${pseudoElement}`, '');

	return {
		normal: {
			label,
			target,
		},
		hover: {
			label: `${label} on hover`,
			target: `${targetWithoutPseudoElement}:hover${
				pseudoElement ? `::${pseudoElement}` : ''
			}`,
		},
		...(targetWithoutPseudoElement !== '' && {
			'canvas hover': {
				label: `${label} on canvas hover`,
				target: `:hover ${targetWithoutPseudoElement.trim()}${
					pseudoElement ? `::${pseudoElement}` : ''
				}`,
			},
		}),
	};
};

export default getNormalAndHoverSelectors;
