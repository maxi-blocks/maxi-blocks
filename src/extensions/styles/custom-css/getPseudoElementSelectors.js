import PSEUDO_ELEMENTS from './constants';

const getPseudoElementSelectors = selectors => {
	return PSEUDO_ELEMENTS.reduce((acc, pseudoElement) => {
		Object.entries(selectors).forEach(([key, selector]) => {
			acc[`${pseudoElement} ${key}`] = {
				label: `${selector.label} ::${pseudoElement}`,
				target: `${selector.target}::${pseudoElement}`,
			};
		});
		return acc;
	}, {});
};

export default getPseudoElementSelectors;
