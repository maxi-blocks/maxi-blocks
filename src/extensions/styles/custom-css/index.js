export { default as createSelectors } from './createSelectors';
