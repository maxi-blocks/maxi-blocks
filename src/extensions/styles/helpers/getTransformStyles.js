/**
 * Internal dependencies
 */
import getLastBreakpointAttribute from '../getLastBreakpointAttribute';
import { validateOriginValue } from '../utils';

/**
 * External dependencies
 */
import { isNumber, isString, isEmpty } from 'lodash';

/**
 * General
 */
const breakpoints = ['general', 'xxl', 'xl', 'l', 'm', 's', 'xs'];

const getTransformStrings = (category, breakpoint, index, obj) => {
	const getLastBreakpointTransformAttribute = ({
		target,
		key,
		hoverSelected,
		keys,
	}) =>
		getLastBreakpointAttribute({
			target,
			breakpoint,
			attributes: obj,
			keys: keys ?? [category, hoverSelected, key],
		});

	const originValueToNumber = value => {
		switch (validateOriginValue(value)) {
			case 'top':
			case 'left':
				return 0;
			case 'middle':
			case 'center':
				return 50;
			case 'bottom':
			case 'right':
				return 100;
			default:
				return value;
		}
	};

	const getScaleString = index => {
		let scaleString = '';
		if (
			index === 'hover' &&
			!getLastBreakpointTransformAttribute({
				target: 'transform-scale',
				keys: [category, 'hover-status'],
			})
		)
			return getScaleString('normal');

		const [x, y] = ['x', 'y'].map(key =>
			getLastBreakpointTransformAttribute({
				target: 'transform-scale',
				key,
				hoverSelected: index,
			})
		);
		if (isNumber(x)) scaleString += `scaleX(${x / 100}) `;
		if (isNumber(y)) scaleString += `scaleY(${y / 100}) `;

		return scaleString;
	};

	const getTranslateString = index => {
		let translateString = '';
		if (
			index === 'hover' &&
			!getLastBreakpointTransformAttribute({
				target: 'transform-translate',
				keys: [category, 'hover-status'],
			})
		)
			return getTranslateString('normal');

		const [x, y, xUnit, yUnit] = ['x', 'y', 'x-unit', 'y-unit'].map(key =>
			getLastBreakpointTransformAttribute({
				target: 'transform-translate',
				key,
				hoverSelected: index,
			})
		);
		if (isNumber(x)) translateString += `translateX(${x}${xUnit ?? '%'}) `;
		if (isNumber(y)) translateString += `translateY(${y}${yUnit ?? '%'}) `;

		return translateString;
	};

	const getRotateString = index => {
		let rotateString = '';
		if (
			index === 'hover' &&
			!getLastBreakpointTransformAttribute({
				target: 'transform-rotate',
				keys: [category, 'hover-status'],
			})
		)
			return getRotateString('normal');

		const [x, y, z] = ['x', 'y', 'z'].map(key =>
			getLastBreakpointTransformAttribute({
				target: 'transform-rotate',
				key,
				hoverSelected: index,
			})
		);

		if (isNumber(x)) rotateString += `rotateX(${x}deg) `;
		if (isNumber(y)) rotateString += `rotateY(${y}deg) `;
		if (isNumber(z)) rotateString += `rotateZ(${z}deg) `;

		return rotateString;
	};

	const getOriginString = index => {
		let originString = '';

		if (
			index === 'hover' &&
			!getLastBreakpointTransformAttribute({
				target: 'transform-origin',
				keys: [category, 'hover-status'],
			})
		)
			return originString;

		const [x, y, xUnit, yUnit] = ['x', 'y', 'x-unit', 'y-unit'].map(key =>
			getLastBreakpointTransformAttribute({
				target: 'transform-origin',
				key,
				hoverSelected: index,
			})
		);

		if (isString(validateOriginValue(x)))
			originString += `${originValueToNumber(x)}% `;
		if (isString(validateOriginValue(y)))
			originString += `${originValueToNumber(y)}% `;

		if (isNumber(validateOriginValue(x)))
			originString += `${x}${xUnit ?? '%'} `;
		if (isNumber(validateOriginValue(y)))
			originString += `${y}${yUnit ?? '%'} `;

		return originString;
	};

	const transformString =
		getScaleString(index) +
		getTranslateString(index) +
		getRotateString(index);
	const transformOriginString = getOriginString(index);

	return [transformString, transformOriginString];
};

const getTransformValue = (obj, category, index) => {
	const response = {};

	breakpoints.forEach(breakpoint => {
		const [transformString, transformOriginString] = getTransformStrings(
			category,
			breakpoint,
			index,
			obj
		);

		const transformObj = {
			...(!isEmpty(transformString) && { transform: transformString }),
			...(!isEmpty(transformOriginString) && {
				'transform-origin': `${transformOriginString}`,
			}),
		};
		if (!isEmpty(transformObj)) response[breakpoint] = transformObj;
	});

	return response;
};

/**
 * Generates size styles object
 *
 * @param {Object} obj Block size properties
 */
const getTransformStyles = (obj, selectors) => {
	const response = {};

	Object.entries(selectors).forEach(([category, targets]) => {
		Object.entries(targets).forEach(([index, targetObj]) => {
			const { target } = targetObj;
			const transformObj = getTransformValue(obj, category, index);
			if (!isEmpty(transformObj))
				response[target] = { transform: transformObj };
		});
	});

	return response;
};

export default getTransformStyles;
