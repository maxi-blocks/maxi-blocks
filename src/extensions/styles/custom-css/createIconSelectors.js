import createSelectors from './createSelectors';

const createIconSelector = (key, selector) => ({
	normal: {
		label: key,
		target: selector,
	},
	svg: {
		label: `${key}'s svg`,
		target: `${selector} svg`,
	},
	insideSvg: {
		label: 'everything inside svg (svg > *)',
		target: `${selector} svg > *`,
	},
	path: {
		label: "svg's path",
		target: `${selector} svg path`,
	},
	hover: {
		label: `${key} on hover`,
		target: `${selector}:hover`,
	},
	hoverSvg: {
		label: `${key}'s svg on hover`,
		target: `${selector}:hover svg`,
	},
	hoverInsideSvg: {
		label: 'everything inside svg on hover (:hover svg > *)',
		target: `${selector}:hover svg > *`,
	},
	hoverPath: {
		label: "svg's path on hover",
		target: `${selector}:hover svg path`,
	},
});

const createIconSelectors = selectors => ({
	...Object.fromEntries(
		Object.entries(selectors).map(([key, selector]) => [
			key,
			createIconSelector(key, selector),
		])
	),
	...createSelectors(selectors, true, true),
});

export default createIconSelectors;
