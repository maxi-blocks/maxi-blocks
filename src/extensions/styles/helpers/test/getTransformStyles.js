import { createSelectors } from '../../custom-css';
import getTransformStyles from '../getTransformStyles';

describe('getTransformStyles', () => {
	const selectors = createSelectors({
		canvas: '',
	});

	it('Get a correct transform styles', () => {
		const object = {
			'transform-origin-general': {
				canvas: {
					normal: {
						x: 'left',
						y: 'top',
						'x-unit': '%',
						'y-unit': '%',
					},
				},
			},
			'transform-rotate-general': {
				canvas: {
					'hover-status': true,
					hover: {
						z: 90,
					},
				},
			},
			'transform-scale-general': {
				canvas: {
					normal: {
						y: 200,
					},
					'hover-status': true,
					hover: {
						x: 200,
					},
				},
			},
			'transform-translate-general': {
				canvas: {
					normal: {
						x: -20,
						y: -10,
						'x-unit': '%',
						'y-unit': '%',
					},
					'hover-status': true,
					hover: {
						x: 20,
						y: 10,
						'x-unit': '%',
						'y-unit': '%',
					},
				},
			},
		};

		const result = getTransformStyles(object, selectors);
		expect(result).toMatchSnapshot();
	});

	it('Get a correct default hover transform styles', () => {
		const object = {
			'transform-origin-general': {
				canvas: {
					'hover-status': true,
					normal: {
						x: 'left',
						y: 'top',
						'x-unit': '%',
						'y-unit': '%',
					},
				},
			},
			'transform-rotate-general': {
				canvas: {
					'hover-status': true,
					normal: {
						x: 30,
						y: 60,
						z: 90,
					},
				},
			},
			'transform-scale-general': {
				canvas: {
					'hover-status': true,
					normal: {
						y: 200,
					},
				},
			},
			'transform-translate-general': {
				canvas: {
					'hover-status': true,
					normal: {
						x: -20,
						y: -10,
						'x-unit': '%',
						'y-unit': '%',
					},
				},
			},
		};

		const result = getTransformStyles(object, selectors);
		expect(result).toMatchSnapshot();
	});
});
