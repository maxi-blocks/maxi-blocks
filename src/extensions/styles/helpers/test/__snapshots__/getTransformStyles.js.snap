// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTransformStyles Get a correct default hover transform styles 1`] = `
{
  "": {
    "transform": {
      "general": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) rotateX(30deg) rotateY(60deg) rotateZ(90deg) ",
        "transform-origin": "0% 0% ",
      },
      "l": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) rotateX(30deg) rotateY(60deg) rotateZ(90deg) ",
        "transform-origin": "0% 0% ",
      },
      "m": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) rotateX(30deg) rotateY(60deg) rotateZ(90deg) ",
        "transform-origin": "0% 0% ",
      },
      "s": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) rotateX(30deg) rotateY(60deg) rotateZ(90deg) ",
        "transform-origin": "0% 0% ",
      },
      "xl": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) rotateX(30deg) rotateY(60deg) rotateZ(90deg) ",
        "transform-origin": "0% 0% ",
      },
      "xs": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) rotateX(30deg) rotateY(60deg) rotateZ(90deg) ",
        "transform-origin": "0% 0% ",
      },
      "xxl": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) rotateX(30deg) rotateY(60deg) rotateZ(90deg) ",
        "transform-origin": "0% 0% ",
      },
    },
  },
}
`;

exports[`getTransformStyles Get a correct transform styles 1`] = `
{
  "": {
    "transform": {
      "general": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) ",
        "transform-origin": "0% 0% ",
      },
      "l": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) ",
        "transform-origin": "0% 0% ",
      },
      "m": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) ",
        "transform-origin": "0% 0% ",
      },
      "s": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) ",
        "transform-origin": "0% 0% ",
      },
      "xl": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) ",
        "transform-origin": "0% 0% ",
      },
      "xs": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) ",
        "transform-origin": "0% 0% ",
      },
      "xxl": {
        "transform": "scaleY(2) translateX(-20%) translateY(-10%) ",
        "transform-origin": "0% 0% ",
      },
    },
  },
  ":hover": {
    "transform": {
      "general": {
        "transform": "scaleX(2) translateX(20%) translateY(10%) rotateZ(90deg) ",
      },
      "l": {
        "transform": "scaleX(2) translateX(20%) translateY(10%) rotateZ(90deg) ",
      },
      "m": {
        "transform": "scaleX(2) translateX(20%) translateY(10%) rotateZ(90deg) ",
      },
      "s": {
        "transform": "scaleX(2) translateX(20%) translateY(10%) rotateZ(90deg) ",
      },
      "xl": {
        "transform": "scaleX(2) translateX(20%) translateY(10%) rotateZ(90deg) ",
      },
      "xs": {
        "transform": "scaleX(2) translateX(20%) translateY(10%) rotateZ(90deg) ",
      },
      "xxl": {
        "transform": "scaleX(2) translateX(20%) translateY(10%) rotateZ(90deg) ",
      },
    },
  },
}
`;
